src/Microsoft.OData.Edm/ExtensionMethods/AnnotationsTargetHelper.cs

```diff
@@ -0,0 +1,205 @@
+
+            if (type != null)
+            {
+                results.Add(type);
+
+                if (targetSegmentsCount - 1 > level)
+                {
+                    return HandleIEdmSchemaType(model, results, type, targetSegments, targetSegmentsCount, level + 1);
+                }
+                else
+                {
+                    // If we have this path MySchema.MyEntityContainer/MyEntitySet/MySchema.MyEntityType
+                    // Should we return null since it's invalid?
+                    return type;
+                }
+            }
+
+            IEdmType edmType = null;
+            IEdmStructuredType structuredType;
+
+            if (edmEntityContainerElement is IEdmEntitySet edmEntitySet)
+            {
+                edmType = edmEntitySet.Type.AsElementType();
+            }
+
+            if (edmEntityContainerElement is IEdmSingleton edmSingleton)
+            {
+                edmType = edmSingleton.Type.AsElementType();
+            }
+
+            if ((structuredType = edmType as IEdmStructuredType) != null)
+            {
+                // .../MyEntitySet/MyProperty/...
+                // .../MyEntitySet/MyNavigationProperty/...
+                // .../MySingleton/MyProperty/...
+                // .../MySingleton/MyNavigationProperty/...
+                IEdmProperty property = structuredType.FindProperty(targetSegments[level]);
+
+                if (property != null)
+                {
+                    results.Add(property);
+
+                    if (targetSegmentsCount - 1 > level)
+                    {
+                        return HandleIEdmProperty(model, results, property, targetSegments, targetSegmentsCount, level + 1);
+                    }
+
+                    return property;
+                }
+            }
+
+            return null;
+        }
+
+        public static IEdmVocabularyAnnotatable HandleIEdmSchemaType(IEdmModel model, List<IEdmElement> results, IEdmSchemaType edmSchemaType, string[] targetSegments, int targetSegmentsCount, int level)
+        {
+            IEdmStructuredType structuredType;
+
+            if ((structuredType = edmSchemaType as IEdmStructuredType) != null)
+            {
+                IEdmProperty property = structuredType.FindProperty(targetSegments[level]);
+
+                if (property != null)
+                {
+                    results.Add(property);
+
+                    if (targetSegmentsCount - 1 > level)
+                    {
+                        return HandleIEdmProperty(model, results, property, targetSegments, targetSegmentsCount, level + 1);
+                    }
+
+                    return property;
+                }
+            }
+
+            return null;
+        }
+
+        public static IEdmVocabularyAnnotatable HandleIEdmProperty(IEdmModel model, List<IEdmElement> results, IEdmProperty edmProperty, string[] targetSegments, int targetSegmentsCount, int level)

```

Comment:
Refer to my other comment https://github.com/OData/odata.net/pull/2769/files#r1382848419